Allow template directory

Pinpickle opened the pull request:
Fixes #2445

If you specify `template_directory` in your theme's `config.yml`, all related template activities will be relative to that directory (which is relative to the theme directory).

It adds `templatespath` to the paths object, and I *think* I managed to change all of the correct instances of `themepath`. This should be checked by someone who knows their stuff.

bobdenotter commented:
Cool! We're waiting for some other PR's to merge (finetuning, and general Travis weirdness), and then I'll look over this and merge it in, if it looks good!

Thanks, again! ^_^

Pinpickle commented:
I thought there were some changes afoot, with all of the failing builds and everything.

Would there be any objection to me removing the depth restriction in the template finder in [TwigExtension](https://github.com/Pinpickle/bolt/blob/issue-2445-template-directory/src/TwigExtension.php#L728)? It would make sense to store templates in subdirectories for logical division and filtering. E.g: people templates under templates/people and post templates under templates/posts and then in your config:

    filter: people/*

GawainLynch commented:
> I thought there were some changes afoot, with all of the failing builds and everything.

That was a misplaced boolean that had @rossriley and I crawling though the git repo forever! Good fun was had by all.

> Would there be any objection to me removing the depth restriction in the template finder [...]

I did want to bring this up with the reset of the team when I replaced the previous `fnmatch()` version of that function...  But it was late beta.  The call on that moves into UI space which I tend to not be too opinionated about as my brain is firmly "nerd" and not really in-line with how most people interact with the application (I personally like the idea though, but note my previous caveat).

GawainLynch commented:
P.S. We actively hang out on FreeNode in #boltcms if ever you want to catch us in person.

GawainLynch commented on src/Configuration/ResourceManager.php:
```diff
@@ -275,6 +275,13 @@ public function initialize()
     {
         $this->setThemePath($this->app['config']->get("general"));
+
+        if (isset($this->app['config']->get('theme')['template_directory'])) {
```
This isn't valid for PHP 5.3, hence the Travis error. Try this:

```php
$theme = $this->app['config']->get('theme');
if (isset($theme['template_directory'])) {
```

Pinpickle commented:
Thanks for the feedback @GawainLynch, I've fixed up that issues and I've gone ahead and added the support for subdirectories. It's completely unobtrusive if you have no templates in subdirectories (they'll just appear as they always have).

![image](https://cloud.githubusercontent.com/assets/3238878/5951180/600c7162-a7c8-11e4-8b86-0c9c7eadccc3.png)

GawainLynch commented:
[just going on picture]

Twig files that start with a '_' should not come up in this list, as they are considered include files (partials).

Pinpickle commented:
Huh, I was getting those before I made any changes, I think due to:

    filter: *.twig

In default contenttypes

Is it desired behaviour to always exclude partials? This can be done by
adding an exclusion name match into the finder, I believe.

Can't look into it now because laptop just died - I will when I get the
chance.

GawainLynch commented:
There might be a recent regression, but on my production boxes it definitely doesn't show partials.

The regex filter `'/^[a-zA-Z0-9]\V+\.twig$/'` should be covering that.

But yes, that is desired behaviour... While backend people might get the purpose of partials, ordinary editors likely won't.

Pinpickle commented:
I definitely still get it in `master`

![image](https://cloud.githubusercontent.com/assets/3238878/5966985/f5d6dc46-a86d-11e4-87b3-3b2ddac04784.png)

This is because when you set a filter in the contenttypes (which is `*.twig` in default `contenttypes.yml`) it overrides the regex you've indicated there. I'll put in an exclusion so the partials are excluded no matter what.

Pinpickle commented:
There we go

![image](https://cloud.githubusercontent.com/assets/3238878/5967214/c17c614e-a86f-11e4-8c33-fd35f54de203.png)

GawainLynch commented:
> I definitely still get it in master

Yeah, I think there might have been an update recently that b0rked it...  No sresss, your fix works and that is the best part! :+1:

Pinpickle commented:
Awesome, happy to have helped!

bobdenotter commented:
MErging in! Thank you again!
